### PR TITLE
[SPARK-25990][SQL] ScriptTransformation should handle different data types correctly

### PR DESCRIPTION
**Pull request by Ngone51**
### What changes were proposed in this pull request?

We should convert Spark InternalRows to hive data via `HiveInspectors.wrapperFor`.

### Why are the changes needed?

We may hit below exception without this change:

```
[info]    org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 1.0 failed 1 times, most recent failure: Lost task 0.0 in stage 1.0 (TID 1, 192.168.1.6, executor driver): java.lang.ClassCastException: org.apache.spark.sql.types.Decimal cannot be cast to org.apache.hadoop.hive.common.type.HiveDecimal
[info]   	at org.apache.hadoop.hive.serde2.objectinspector.primitive.JavaHiveDecimalObjectInspector.getPrimitiveJavaObject(JavaHiveDecimalObjectInspector.java:55)
[info]   	at org.apache.hadoop.hive.serde2.lazy.LazyUtils.writePrimitiveUTF8(LazyUtils.java:321)
[info]   	at org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe.serialize(LazySimpleSerDe.java:292)
[info]   	at org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe.serializeField(LazySimpleSerDe.java:247)
[info]   	at org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe.doSerialize(LazySimpleSerDe.java:231)
[info]   	at org.apache.hadoop.hive.serde2.AbstractEncodingAwareSerDe.serialize(AbstractEncodingAwareSerDe.java:55)
[info]   	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread.$anonfun$run$2(ScriptTransformationExec.scala:300)
[info]   	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread.$anonfun$run$2$adapted(ScriptTransformationExec.scala:281)
[info]   	at scala.collection.Iterator.foreach(Iterator.scala:941)
[info]   	at scala.collection.Iterator.foreach$(Iterator.scala:941)
[info]   	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
[info]   	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread.$anonfun$run$1(ScriptTransformationExec.scala:281)
[info]   	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   	at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1932)
[info]   	at org.apache.spark.sql.hive.execution.ScriptTransformationWriterThread.run(ScriptTransformationExec.scala:270)
```

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Added new test. But please note that this test returns different result between Hive1.2 and Hive2.3 due to `HiveDecimal` or `SerDe` difference(don't know the root cause yet).

### CODE REVIEW

**Comment by Ngone51** [comment]
cc @cloud-fan

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ScriptTransformationSuite.scala [hunk 1]
```diff
@@ -186,6 +191,40 @@ class ScriptTransformationSuite extends SparkPlanTest with TestHiveSingleton wit
...
+
+      val query = sql(
+        s"""
+          |SELECT
+          |TRANSFORM(a, b, c, d, e)
+          |USING 'python $scriptFilePath' AS (a, b, c, d, e)
+          |FROM v
+        """.stripMargin)
+
+      val castDecimal: Column => Column = if (HiveUtils.isHive23) {
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ScriptTransformationSuite.scala]: can we add some code comments to explain it?

Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ScriptTransformationSuite.scala]: and maybe name it as `decimalToString`

**Review by cloud-fan — APPROVED** [review]

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ScriptTransformationSuite.scala [hunk 2]
```diff
@@ -186,6 +191,43 @@ class ScriptTransformationSuite extends SparkPlanTest with TestHiveSingleton wit
...
+      val query = sql(
+        s"""
+          |SELECT
+          |TRANSFORM(a, b, c, d, e)
+          |USING 'python $scriptFilePath' AS (a, b, c, d, e)
+          |FROM v
+        """.stripMargin)
+
+      // In Hive1.2, it does not do well on Decimal conversion. For example, in this case,
+      // it converts a decimal value's type from Decimal(38, 18) to Decimal(1, 0). So we need
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ScriptTransformationSuite.scala]: This is a little confusing, if Hive 1.2 uses `Decimal(1, 0)` as the type, then `cast("decimal(1, 0)")` should be a no-op?

Comment (cloud-fan) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/ScriptTransformationSuite.scala]: If we don't add `cast("decimal(1, 0)")`, what's the result of hive 1.2 and 2.3?

**Review by cloud-fan — APPROVED** [review]

**Comment by Ngone51** [comment]
Jenkins, retest this please.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.0!

**Comment by AngersZhuuuu** [comment]
@Ngone51 
Dose your pr make transform support array/map datatype as script's input column?

**Comment by Ngone51** [comment]
> Dose your pr make transform support array/map datatype as script's input column?

@AngersZhuuuu No. This doesn't add support for new types but just improve already supported types.

**Pull request closed without merging**